[docs] Add persisted-queries command (#1698)
apollographql/apollo-cli: docs/source/commands/persisted-queries.mdx

---
title: The Rover persisted-queries command
description: Publish persisted query manifests to your persisted queries list
minVersion: 0.17.2
---

import AuthNotice from '../../shared/auth-notice.mdx';

> With [GraphOS Enterprise](/graphos/enterprise/), you can enhance a graph's security by maintaining a **persisted query list** (**PQL**) for your graph's self-hosted router. The [Apollo Router](/router/) checks incoming requests against the PQL, a safelist of trusted operations made by your first-party apps.

Reviewer: kind of a small nit here and I don't love my wording change but I want to communicate here that you can use PQLs without enabling full safelisting in the router. what do you think?

```suggestion
With [GraphOS Enterprise](/graphos/enterprise/), you can enhance a graph's security by maintaining a **persisted query list** (**PQL**) for your graph's self-hosted router. The [Apollo Router](/router/) checks incoming requests against the PQL, and can be configured to reject operations that are not in the list.
```

Preregistering trusted operations to a PQL has two steps:

1. Generating persisted queries manifests (PQM) using client-specific tooling
2. Publishing PQMs to the PQL using Rover

This article details the latter. For more information on PQM generation and other implementation steps, see the [GraphOS persisted queries documentation](/graphos/operations/persisted-queries).

## Publishing a PQM

### `persisted-queries publish`

<AuthNotice />

You can use Rover to publish a PQM for any client to any [existing PQL](/graphos/operations/persisted-queries#1-pql-creation-and-linking). 

Run the `persisted-queries publish` command, like so:

```bash
rover persisted-queries publish my-graph@my-variant \
  --manifest ./persisted-query-manifest.json
```

- The `my-graph@my-variant` argument is the [`graph ref`](../conventions#graph-refs) of _any_ variant the PQL is [linked to](/graphos/operations/persisted-queries#12-link-the-pql-to-variants).
    - Graph refs have the format `graph-id@variant-name`.
- Use the `--manifest` option to provide the path to the manifest you want to publish.

> Instead of `my-graph@my-variant`, you can also use the `--graph-id` and `--list-id` options to specify the PQL you want to publish to:

Reviewer: ```suggestion
Instead of `my-graph@my-variant`, you can combine the `--graph-id` and `--list-id` options to specify the PQL you want to publish to:
```

i think using `combine` makes it clearer that both arguments are required in this command form

```bash
rover persisted-queries publish \
  --graph-id my_graph --list-id dc4b4040-30fc-4bd1-94a3-5fc1c722acc9 \
  --manifest ./persisted-query-manifest.json
```

- Use `--graph-id` to provide the graph's ID
- Use `--list-id` to provide the PQL's UUID

**The `persisted-queries publish` command does the following:**

> 1. It publishes all operations in the provided manifest file to the PQL linked to the specified variant, or to the specified PQL.

Reviewer: ```suggestion
1. Publishes all operations in the provided manifest file to the PQL linked to the specified variant, or to the specified PQL.
```

    - Publishing a manifest to a PQL is additive. Any _existing_ entries in the PQL remain.
    - If you publish an operation with the same `id` but different details from an existing entry in the PQL, the entire publish command fails with an error.

> 2. It updates any _other_ variants that the PQL is applied to so that routers associated with those variants can fetch their updated PQL.

Reviewer: ```suggestion
2. Updates any _other_ variants that the PQL is applied to so that routers associated with those variants can fetch their updated PQL.
```
